step-templates/clone-tenant.json

```diff
@@ -3,14 +3,24 @@
to Clone' was not found. Please provide one and try again.\"\n}\n\nif ([string]::IsNullOrWhiteSpace($tenantName)) {\n    throw \"The step parameter 'New Tenant Name' was not found. Please provide one and try again.\"\n}\n\nfunction Invoke-OctopusApi {\n    param(\n        [Parameter(Position = 0, Mandatory)]$Uri,\n        [ValidateSet(\"Get\", \"Post\", \"Put\", \"Delete\")]$Method = 'Get',\n        $Body\n    )\n    \n    $uriParts = @($octopusBaseUrl, $Uri.TrimStart('/'))    \n    $uri = ($uriParts -join '/')\n\n    Write-Verbose \"Uri: $uri\"\n    \n    $requestParameters = @{\n        Uri = $uri\n        Method = $Method\n        Headers = @{ \"X-Octopus-ApiKey\" = $apiKey }\n        UseBasicParsing = $true\n    }\n    \n    if ($null -ne $Body) { $requestParameters.Add('Body', ($Body | ConvertTo-Json -Depth 10)) }\n    \n    return Invoke-WebRequest @requestParameters | % Content | ConvertFrom-Json\n}\n\nfunction Test-SpacesApi {\n\tWrite-Verbose \"Checking API compatibility\";\n\t$rootDocument = Invoke-OctopusApi 'api/';\n    if($rootDocument.Links -ne $null -and $rootDocument.Links.Spaces -ne $null) {\n    \tWrite-Verbose \"Spaces API found\"\n    \treturn $true;\n    }\n    Write-Verbose \"Pre-spaces API found\"\n    return $false;\n}\n\nif([string]::IsNullOrWhiteSpace($spaceId)) {\n\tif(Test-SpacesApi) {\n      \t$spaceId = $OctopusParameters['Octopus.Space.Id'];\n      \tif([string]::IsNullOrWhiteSpace($spaceId)) {\n          \tthrow \"This step needs to be run in a context that provides a value for the 'Octopus.Space.Id' system variable. In this case, we received a blank value, which isn't expected - please reach out to our support team at https://help.octopus.com if you encounter this error or try providing the Space Id parameter.\";\n      \t}\n\t}\n}\n\n$apiPrefix = \"api/\"\n$tenantUrlBase = @($octopusBaseUrl, 'app#')\n\nif ($spaceId) {\n\t$apiPrefix += $spaceId\n    $tenantUrlBase += $spaceId\n}\n\nWrite-Host \"Fetching source tenant\"\n$tenant = Invoke-OctopusApi \"$apiPrefix/tenants/$tenantToClone\"\n\n$sourceTenantId = $tenant.Id\n$sourceTenantName = $tenant.Name\n$tenant.Id = $null\n$tenant.Name = $tenantName\n\nif ($cloneTags -ne $true) {\n\tWrite-Host \"Clearing tenant tags\"\n    $tenant.TenantTags = @()\n}\n\nWrite-Host \"Creating new tenant\"\n$newTenant = Invoke-OctopusApi \"$apiPrefix/tenants\" -Method Post -Body $tenant\n\nif ($cloneVariables -eq $true) {\n\tWrite-Host \"Cloning variables\"\n    $variables = Invoke-OctopusApi $tenant.Links.Variables\n    $variables.TenantId = $newTenant.Id\n    $variables.TenantName = $tenantName\n\n    Invoke-OctopusApi $newTenant.Links.Variables -Method Put -Body $variables | Out-Null\n}\n\n$tenantUrl = ($tenantUrlBase + \"tenants\" + $newTenant.Id + \"overview\") -join '/'\n$sourceTenantUrl = ($tenantUrlBase + \"tenants\" + $sourceTenantId + \"overview\") -join '/'\n\nWrite-Highlight \"New tenant [$tenantName]($tenantUrl) has been cloned from [$sourceTenantName]($sourceTenantUrl)\""
```

Comment:
This is a great start! 

Setting the `SecurityProtocol` in this way is a bit of a trap: it will make the code _only_ support TLS 1.2 

Have a look at how the client does this [here](https://github.com/OctopusDeploy/OctopusClients/blob/c5edcce910bdc8be2051d201bba5b9d4892bc2e6/source/Octopus.Cli/CliProgram.cs#L25) There is similar code in Server that does the same thing.

Yup - that means that when TLS 1.3 comes out, and people start wanting that, we'll have to update these values... thankfully MS is moving away from this design decision in .NET core.

While I couldn't find examples of other library steps using the Octopus.Client.dll library (and there may be a good reason for that - they possibly run in locations where this is not available?), doing so would take care of this for you - as well as testing for spaces compatibility etc. Not suggesting that you _must_ do that, but perhaps its a possibility.

Reply:
Oh, interesting! I'll make the adjustment tomorrow and resubmit.

I think that other steps don't use Octopus.Client.dll for API calls because that means it'll need to be installed on the machine running the step - which isn't necessarily a bad thing, but means that extra configuration would be required for the step to work.

Reply:
I looked at using Octopus.Client instead of web requests but it started to turn into a deeper rabbit hole than expected. Instead, I updated both steps to match Octopus.Client's logic.

```powershell
$securityProtocol = [Net.SecurityProtocolType]::Tls -bor [Net.SecurityProtocolType]::Tls11 -bor [Net.SecurityProtocolType]::Tls12
[Net.ServicePointManager]::SecurityProtocol = $securityProtocol
```

Reply:
Hmm, maybe hold on this. In a different template, I'm getting weird issues with the checkbox parameters that I didn't see in these. I want to make sure I didn't miss anything.

Reply:
Just a mistake in my other template, these are good 👍

Reply:
>  deeper rabbit hole
Totally understand. The above changes to the TLS flags look good.